link-grammar/parse/prune.c

```diff
@@ -1129,17 +1129,169 @@ static int pp_prune(Sentence sent, Parse_Options opts)
+		if ((r_nearest_word[w] > sent->length-1) ||
+			 (r_nearest_word[w] < w))
+		{
+			prt_error("Warning: w[%d].nearest_word_right %d\n",
+			          w, r_nearest_word[w]);
+			continue;
+		}
+		if (l_nearest_word[w] > w)
+		{
+			prt_error("Warning: w[%d].nearest_word_left %d\n",
+			          w, l_nearest_word[w]);
+			continue;
+		}
+
+		if (sent->word[w].optional) continue;
+
+		/* Remove disjuncts that are blocked by l->r links. */
+		for (WordIdx rw = w+1; rw < r_nearest_word[w]; rw++)
+		{
+			if (sent->word[rw].optional) continue;
+
+			for (Disjunct **dd = &sent->word[rw].d; *dd != NULL; /* See: NEXT */)
+			{
+				Disjunct *d = *dd; /* just for convenience */
+
+				if ((NULL != d->left) && d->left->nearest_word < w)
+				{
+#if 0
+					printf("w %d nw_right %d rw %d nw_left %d dl %p:",
+					       w, r_nearest_word[w], rw, d->left->nearest_word, d);
+					print_connector_list(d->left);
+					printf("\n");
+#endif
+
+					/* Discard the current disjunct. */
+					*dd = d->next; /* NEXT - set current disjunct to the next one */
+					d->next = free_later;
+					free_later = d;
+					total_deleted++;
+				}
+				else
+				{
+					dd = &d->next; /* NEXT */
+				}
+			}
+		}
+
+		/* Remove disjuncts that are blocked by mandatory r->l links. */
+		for (WordIdx lw = l_nearest_word[w]+1; lw < w; lw++)
+		{
+			if (sent->word[lw].optional) continue;
+
+			for (Disjunct **dd = &sent->word[lw].d; *dd != NULL; /* See: NEXT */)
+
+			{
+				Disjunct *d = *dd; /* just for convenience */
+
+				if ((NULL != d->right) && d->right->nearest_word > w)
+				{
+#if 0
+					printf("w %zu nw_left %d rw %zu nw_right %d dl %p:",
+					       w, l_nearest_word[w], lw, d->right->nearest_word, d);
+					print_connector_list(d->right);
+					printf("\n");
+#endif
+
+					/* Discard the current disjunct. */
+					*dd = d->next; /* NEXT - set current disjunct to the next one */
+					d->next = free_later;
+					free_later = d;
+					total_deleted++;
+				}
+				else
+				{
+					dd = &d->next; /* NEXT */
+				}
+			}
+		}
+	}
+
+	free_disjuncts(free_later);
+	lgdebug(D_PRUNE, "Debug: cross_links_prune removed %d\n", total_deleted);
+	return total_deleted;
+}
 
 /**
- * Do the following pruning steps until nothing happens:
- * power pp power pp power pp....
- * Make sure you do them both at least once.
+ * Performs the 3 pruning steps.
+ * Notes:
+ * - No benefit was found to perform pp_prune after cross_links_prune.
+ * - If cross_links_prune deleted even few disjuncts, power_prune will
+ *   most probably may delete many more.
+ * - Empirically it was found that if power_prune deleted a few tens of
+ *   disjuncts, than cross_links_prune has a good chance to delete again a
+ *   significant number.
  */
 void pp_and_power_prune(Sentence sent, Parse_Options opts)
 {
 	power_prune(sent, opts);
 	pp_prune(sent, opts);
 
+	int i;
+	do
+	{
+		i = cross_links_prune(sent, opts);
+		if (i == 0) break;
+		i = power_prune(sent, opts);
+	}
+	while (i > 25);
```

Comment:
It could be` i>0`, since `cross_links_prune()` is very lightweight.
`25` was based on a printout showing that only rarely `power_prune()` does a useful work when `i<20`. I will check that again when I add more constraint that are able to discard disjuncts.

Reply:
> 25 was based on a printout showing that only rarely power_prune() does a useful work 

25 was based on a printout showing that only rarely ~power_prune()~ `cross_links_prune()`  does a useful work